### ISSUE TITLE
What is this line for? Nothing happens when I delete it.

### ISSUE THREAD

**Issue by frdteknikelektro** [thedevdojo/voyager#2600]
I have a problem like this. https://github.com/laravel/framework/issues/23014

https://github.com/the-control-group/voyager/blob/e9c530f0f9406e09107b3113a3232f97e0bb7230/publishable/database/seeds/MenuItemsTableSeeder.php#L17

**Comment by emptynick** [comment]
This is definitely not enough information for us to help you

**Comment by frdteknikelektro** [comment]
Okay, the problem is happen when I make phpunit test and running seeder.
The Test failed to recognize controller route, it fail to detect controller namespace,
I've been to uncomment some seeder and come to MenuItemsTableSeeder.
I found that seeder is executing `require base_path('routes/web.php')`.
And all the logic become wrong because `routes/web.php` file should be included like this.
```
Route::middleware('web')
                 ->namespace('App\Http\Controllers')
                 ->group(base_path('routes/web.php'));
```

That's why I'm asking what that line is use for?

**Comment by fletch3555** [comment]
The line is there because routes need to be loaded before `MenuItems` can be loaded, since they use dynamic routes.  The one that Voyager ships with works fine, so I'm assuming you've added custom entries to the `MenuItemsTableSeeder`.

I agree there is probably a better way to load the routes to handle this case.  I'm open to Pull Requests for it.  I would suggest something along the lines of `app(RouteServiceProvider::class)->map()` in case something else was changed

**Comment by nickbryan** [comment]
HI @frdteknikelektro, I am having this exact same issue and its killing me. I was about to create an issue today. I'll add some info here that may help.

I am creating some feature tests that hit an endpoint in my custom controller.
```php
    public function testDevelopersCanSelectACompanyToView()
    {
        $company = factory(Company::class)->create();
        factory(CompanyHierarchy::class, 5)->create([
            'company_id' => $company->id,
            'type_id' => factory(CompanyHierarchyType::class)->create(['company_id' => $company->id])->id
        ]);

        $user = factory(User::class)->states(User::ROLE_WRANX_DEVELOPER)->create();
        $response = $this->actingAs($user)->get('/company_hierarchy/show?company=' . $company->name);
        $response->assertRedirect('/company_hierarchy?company_id=' . $company->id);

        $response = $this->followRedirects($response);
        $userHierarchy = CompanyHierarchy::where('company_id', '=', $company->id)->with('type')->get();
        $response->assertViewHas('items', $userHierarchy);
    }
````

The route is defined as such
```php
    Voyager::routes();

    Route::get('/company_hierarchy/show', [
        'uses' => 'CompanyHierarchyController@show',
        'as' => 'company_hierarchy_show'
    ]);
```

My data type entry is defined as
```php
    $dataType = $this->dataType('slug', 'company_hierarchy');
        if (!$dataType->exists) {
            $dataType->fill([
                'name'                  => 'company_hierarchy',
                'display_name_singular' => 'Company Hierarchy',
                'display_name_plural'   => 'Companies Hierarchies',
                'icon'                  => 'voyager-shop',
                'model_name'            => 'App\\Models\\CompanyHierarchy',
                'policy_name'           => 'App\\Policies\\CompanyHierarchyPolicy',
                'controller'            => 'App\\Http\\Controllers\\CompanyHierarchyController',
                'generate_permissions'  => 1,
                'description'           => '',
                'server_side'           => true,
            ])->save();
        }
```

The 500 error I am getting states `Class CompanyHierarchyController does not exist`.

If I specify the FQN for the controller in the route it works fine but i must add a `\` to the start of the FQN otherwise the namespace when viewing outside of test will be `App\Http\Controllers\App\Http\Controllers\CompanyHierarchyController`. I have got a temp fix for this in place by adding a route group around all my Voyager route overrides.

```php
Route::group(['namespace' => '\App\Http\Controllers'], function() {
    Route::get('/company_hierarchy/show', [
        'uses' => 'CompanyHierarchyController@show',
        'as' => 'company_hierarchy_show'
    ]);
```

My controller method looks like
```php
    public function show(Request $request): RedirectResponse
    {
        $this->authorize('view-all', CompanyHierarchy::class);

        if ($request->has('company')) {
            $company = Company::where('name', '=', $request->input('company'))->first();

            if ($company) {
                return redirect()->route('voyager.company_hierarchy.index', ['company_id' => $company->id]);
            }
        }

        return redirect()->route('voyager.company_hierarchy.index')->with([
            'message'   => 'Company not found for ' . $request->input('company'),
            'alert-type' => 'error',
        ]);
    }
```

The setup method for my tests is
```php
class CompanyHierarchyTest extends TestCase
{
    use RefreshDatabase;

    public function setUp()
    {
        parent::setUp();

        Artisan::call('db:seed', ['--class' => VoyagerDatabaseSeeder::class]);
    }
```

I hope this provides some more information for you to resolve this issue.

@fletch3555 let me know if you need anymore information on this.

**Comment by nickbryan** [comment]
Is it not possible to load the routes in the same way as the `RouteServiceProvider` in the seeder?

```php
Route::middleware('web')
             ->namespace($this->namespace)
             ->group(base_path('routes/web.php'));
```

I will try test this today and get back to you.

**Comment by nickbryan** [comment]
Hi guys, so this seems to have worked anyway.

I have replaced
```php
if (file_exists(base_path('routes/web.php'))) {
            require base_path('routes/web.php');
```

With
```php
if (file_exists(base_path('routes/web.php'))) {
            Route::middleware('web')->namespace('App\Http\Controllers')->group(base_path('routes/web.php'));
```

Everything seems to be working as expected now. Tests are all passing and seeds working as expected.

**Comment by frdteknikelektro** [comment]
Yes that is also my workaround.

And if search in the Voyager package there is a lot of
`require base_path('routes/web.php')`

**Comment by nickbryan** [comment]
@fletch3555 do you see any issues with this?

**Comment by fletch3555** [comment]
Yes, What happens when someone changes the app name (`artisan app:name`)? We would then have a hard-coded namespace in our package tying back to the "App", which is _no bueno_

Instead, we should use what I suggested in my last comment in this thread (10 days ago).  Please test that and see if it works.

**Comment by nickbryan** [comment]
Hi @fletch3555, thanks for the reply, totally missed your suggestion there, must have skimmed through when reading this. I will test that out and get back to you.

**Comment by nickbryan** [comment]
So i tried this out this morning and getting `Illuminate\Contracts\Container\BindingResolutionException: Unresolvable dependency resolving [Parameter #0 [ <required> $app ]] in class Illuminate\Support\ServiceProvider` when trying to call your code, shouldnt the container be resolving this dependency anyway? Confused

**Comment by nickbryan** [comment]
Ah @fletch3555 its because the $app dep in `Illuminate\Support\ServiceProvider` constructor is not type hinted. This means the ioc container can not properly resolve the dependency. Adding this to the `RouteServiceProvider` fixes the issue but its not idea for putting this in your code base i guess.

```php
    public function __construct(Application $app)
    {
        parent::__construct($app);
    }
```

Alternatively this works perfectly fine:
```php
app(RouteServiceProvider::class, ['app' => app(\Illuminate\Contracts\Foundation\Application::class)])->map(); 
```

**Comment by vblinden** [comment]
Can't we just do the following? Cause it only need the Voyager routes right?

```php
Route::group(['prefix' => 'admin'], function () {
    Voyager::routes();
});
```

All tests are still running fine on my system.

**Comment by emptynick** [comment]
This was "fixed" in #2792
